File: templates/ts/api-plugin-from-scratch-oauth/src/functions/middleware/authMiddleware.ts
import { HttpRequest } from "@azure/functions";
import { TokenValidator } from "./tokenValidator";
import config from "./config";
import { getEntraJwksUri, CloudType } from "./utils";

/**
 * Middleware function to handle authorization using JWT.
 *
 * @param {HttpRequest} req - The HTTP request.
 * @returns {Promise<boolean>} - A promise that resolves to a boolean value.
 */
export async function authMiddleware(req?: HttpRequest): Promise<boolean> {
  // Get the token from the request headers
  const token = req.headers.get("authorization")?.split(" ")[1];
  if (!token) {
    return false;
  }

  try {
    // Get the JWKS URL for the Microsoft Entra common tenant
    const entraJwksUri = await getEntraJwksUri(config.aadAppTenantId, CloudType.Public);

    // Create a new token validator with the JWKS URL
    const validator = new TokenValidator({
      jwksUri: entraJwksUri,
    });

    const options = {
      allowedTenants: [config.aadAppTenantId],
      audience: config.aadAppClientId,
      issuer: `https://login.microsoftonline.com/${config.aadAppTenantId}/v2.0`,
      scp: ["repairs_read"],
    };
    // Validate the token
    await validator.validateToken(token, options);

    return true;
  } catch (err) {
    // Handle JWT verification errors
    console.error("Token is invalid:", err);
    return false;
  }
}

File: templates/csharp/api-plugin-from-scratch-oauth/Middleware/AuthMiddleware.cs
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.Extensions.Logging;

public class AuthMiddleware
{
    private readonly TokenValidator _tokenValidator;
    private readonly string _audience;
    private readonly string _issuer;
    private readonly string[] _allowedTenants;
    private readonly string[] _scopes;

    public AuthMiddleware(TokenValidator tokenValidator, string audience, string issuer, string[] allowedTenants, string[] scopes)
    {
        _tokenValidator = tokenValidator;
        _audience = audience;
        _issuer = issuer;
        _allowedTenants = allowedTenants;
        _scopes = scopes;
    }

    public async Task<bool> ValidateTokenAsync(HttpRequestData req, ILogger logger)
    {
        if (!req.Headers.TryGetValues("Authorization", out var authHeaders))
        {
            return false;
        }

        var token = authHeaders.FirstOrDefault()?.Split(" ").Last();
        if (string.IsNullOrEmpty(token))
        {
            return false;
        }

        try
        {
            var options = new TokenValidationOptions
            {
                AllowedTenants = _allowedTenants,
                Audience = _audience,
                Issuer = _issuer,
                Scopes = _scopes,
            };
            await _tokenValidator.ValidateTokenAsync(token, options);
            return true;
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Token is invalid");
            return false;
        }
    }
}